Task: Create Plans page for Jetpack App Site Creation

staskus opened the pull request:
Fixes: #82131

Investigation: https://github.com/wordpress-mobile/WordPress-iOS/issues/21598#issuecomment-1731081950
Discussion: p1695414275007579-slack-CFFF01Q4V
More context: pcdRpT-3RO-p2, 

## Requirements

- Free, Personal, Premium, Business, E-commerce Plans
- Input: A domain name
- Output: Selected plan

## Description

Created Plans page for Jetpack App Site Creation. It is intended to be reused and integrated into the native app flows.

## Added routes:

### For mobile-app user agent:
- `jetpack-app` redirects to `jetpack-app/plans`
- `jetpack-app/plans`

### Other cases:
- `jetpack-app` redirects home
- `jetpack-app/plans` redirects home

## Tasks

#### Basic setup

- [x] Created jetpack-app-plans directory
- [x] Added an entry into sections.js with a path and module
- [x] Created controller.jsx that displays JetpackAppPlans component and passes query parameters
- [x] Created index.js defining a page and including jetpackAppPlans controller
- [x] Created main.jsx that shows a loading indicator while querying plans, and PlansFeaturesMain component once plans are loaded
- [x] Created style.scss with jetpack-app-plans style that includes plan-features-layout-switcher to support media queries when displaying plans grid

#### Functionality

- [x] Setup specific plan selection for the page. Pass `intent` props to `PlansFeatureMain`. Consider creating a similar `intent` to "plans-jetpack-app" but for a site creation, which includes free plan as well 
- [x] Setup required query parameter for the feature in `controller.jsx`that would satisfy the requirements
- [x] Handle plan selection (`onUpgradeClick`) to pass `cartItem` for a mobile application once the plan is selected
- [x] Consider handling `removePaidDomain` which is triggered when paid domain is set and free plan is selected
...

#### Style

- [x] Hide action for "yearly/monthly" interval switch
- [x] Hide site switcher that is displayed on iPad-size page
- [x] Finish setting up `style.scss` which would include all the relevant styles so the page would appear fine for phones, and tablets. See styles of other features that are using `PlansFeaturesMain`
- [x] Show "Choose the perfect plan" title, the same one shown in post-signup plans flow

#### Coding Standards

- [x] Make sure [New Component Checklist](https://github.com/Automattic/wp-calypso/blob/trunk/docs/new-component-checklist.md) is followed
- [ ] Tracking

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Testing Instructions

### UI
1.  Open `.com/jetpack-app-plans`
2. Confirm Plans selection page is opened
3. Confirm Free, Personal, Premium, Business, and E-Commerce Plans are displayed
4. Confirm there're no headers and sidebars
5. Confirm the page is responsive to viewport and orientation changes

### paid_domain_name parameter
1. Pass `paid_domain_name` parameter (`.com/jetpack-app/plans?paid_domain_name=greatdomain.com`)
2. Confirm tapping on `Free` opens a pop-up
3. Confirm domain name is visible in the header text as well as next to the plans
4. Confirm tapping on a plan redirects to the same page with appended `plan_id` and `plan_slug` parameter. The app can use this redirection to come back to the native flow.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

staskus committed:
Create a basic structure for /jetpack-app-plans page

- Created jetpack-app-plans directory
- Added an entry into sections.js with a path and module
- Created controller.jsx that displays JetpackAppPlans component and passes query parameters
- Created index.js defining a page and including jetpackAppPlans controller
- Created main.jsx that shows a loading indicator while querying plans, and PlansFeaturesMain component once plans are loaded
- Created style.scss with jetpack-app-plans style that includes plan-features-layout-switcher to support media queries when displaying plans grid

staskus commented:
## Changelog

I created a basic structure for `/jetpack-app-plans` page following [documentation guide for creating a new section](https://github.com/Automattic/wp-calypso/blob/trunk/docs/guide/hello-world.md).  I used other features that include `PlanFeaturesMain` component as an example.

https://github.com/Automattic/wp-calypso/assets/4062343/bd899a2d-519f-41be-8856-3fed83db1d92

Now, we can:
- Open `/jetpack-app-plans` that shows a basic plans grid
- Pass `domainName` parameter (`/jetpack-app-plans?domainName=example.com`) to display the domain on the cards and display a popup when free domain is selected
- Handle `onUpgradeClick` and `onUpgradeClick` events

I added tasks to the description. I think the most important next step is to make sure we can set the required inputs, and get the required outputs from and into our app to support the requirements. There will be more issues and caveats that you may find, please include them into tasks and/or mention in the changelog.

CC: @guarani @ravishanker

Commit:
Remove use of deprecated params

guarani commented:
> I created a basic structure for `/jetpack-app-plans` page following [documentation guide for creating a new section](https://github.com/Automattic/wp-calypso/blob/trunk/docs/guide/hello-world.md). I used other features that include `PlanFeaturesMain` component as an example.

Good find! This is working great, thanks for making this happen @staskus! The code looks good and I only removed some parameters from `PlansFeaturesMain` that are [marked deprecated](https://github.com/Automattic/wp-calypso/blob/3ac052541fc3bec938523a31c6b17a8dc849d5f5/client/my-sites/plans-features-main/index.tsx#L151-L152) and re-tested. 

I haven't found the time today to add anything else to this PR yet.

staskus committed:
Set intent for Jetpack app plans in site creation

staskus committed:
Hide yearly/monthly selector

staskus committed:
Pass selected plan through redirection url
- Allow redirect_to parameter to URL which is then used to callback with a selected plan
- Do not pass selected plan if free plan is selected

staskus committed:
Use snake case for passing cart item

staskus commented:
## Changelog

✅  Configured plan selection to include free, personal, premium, business, and e-commerce (https://github.com/Automattic/wp-calypso/pull/82135/commits/c5cfdc1fec9301789be4382309b578c54bd714dc
✅ Hid the yearly/monthly selector (https://github.com/Automattic/wp-calypso/pull/82135/commits/377e11f6cd290dfdf2c61c0b5fed28995afcb623)
✅ Handle onUpgradeClick and removePaidDomain events (https://github.com/Automattic/wp-calypso/pull/82135/commits/3c896d613927153e1013da96da843f88bed3a472)

### More details on handling onUpgradeClick and removePaidDomain events

The goal was to find out the way to pass information from the page into the app. I wanted to avoid calling platform-specific code and keep things as simple as possible. 

My solution is to use `window.location.href` API that can be observed on web views and handled appropriately. 

#### Example

1. We pass `redirect_to` parameter `wp.com/jetpack-app-plans?domain_name="somedomain.com"&redirect_to="jetpackapp://plans`
2. When a user taps on a personal plan, we call `window.location.href="jetpackapp://plans?cart_item={product_slug: value_bundle, product_id: 1003}` (encoded)
3. Apps can handle it using usual redirection observing:

```
    override func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        let url = navigationAction.request.url
        /// Depends on what "redirect_to" parameter we pass
        if url?.scheme == "jetpackapp" {
            if let url = url, let components = URLComponents(url: url, resolvingAgainstBaseURL: true), let cartItem = components.queryItems?.first?.value {
                print(cartItem) // {"product_id":1009,"product_slug":"personal-bundle"}
            }
          decisionHandler(.cancel)
          return
        }
        decisionHandler(.allow)
    }
```
4. Then we can use existing cart APIs to add both domain and product to a cart after the site is created (https://github.com/wordpress-mobile/WordPress-iOS/issues/21636)

#### Printing `window.location.href` into console video

https://github.com/Automattic/wp-calypso/assets/4062343/38683b55-861a-46c0-958c-b6e86247da1c

#### Using a dummy plan selection in the site creation flow that sets `window.location.href` value and adds it to cart video

https://github.com/Automattic/wp-calypso/assets/4062343/103dbdf1-a6b4-4d5e-9aba-d6d0a608fd4a

### Issues

I couldn't identify a way to test these Calypso changes on the app, our authentication methods for web views don't work with local Calypso builds. Therefore I resorted to mocking the outputs from within the app with a dummy `HTML` page to confirm that it's a valid way to observe and get the selected plan.

CC: @guarani @ravishanker

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation

Commit:
Fix plan retrieval from cart

Use getPlanCartItem to get plan

guarani commented:
👋 There's an error in the console:
```
Cannot read properties of undefined (reading 'getProductId')
```

https://github.com/Automattic/wp-calypso/assets/1898325/e480967a-aa22-428a-a9a3-2e6b65fc1354

Based on your screen recordings above, there was no error, and I'm not sure why it wasn't happening before and is happening now. Please let me know if I'm missing something. `onUpgradeClick` returns an array of plan items, so it has no `getProductId` method. I used `getPlanCartItem` to get the plan item from the array (there should only be one), and get the product ID and slug from there.

staskus commented:
> `onUpgradeClick` returns an array of plan items, so it has no `getProductId` method. I used `getPlanCartItem` to get the plan item from the array (there should only be one), and get the product ID and slug from there.

Interesting. `onUpgradeClick` returned to me a `MinimalRequestCartProduct` which only contained `product_slug`. O used this slug to get a full product and then take `productId` out of it. I'll look further at what could be the issue 🤔

staskus committed:
Check if cartItems exist before fetching plan

staskus committed:
Hide top and sidebars in jetpack-app-plans flow

staskus committed:
Add a header to jetpack-app-plans page

staskus committed:
Refactor to improve code style

staskus committed:
Show a different header label when no domain name is passed

staskus committed:
Hide feature comparison

staskus committed:
Rename domain_name to paid_domain_name for more clarity

staskus committed:
Reuse the same translation from previous label

staskus commented:
## Changelog

✅ Hid main header and sidebar
✅ Added page header with titles
✅ Other refactorings and improvements to style and localization

https://github.com/Automattic/wp-calypso/assets/4062343/b2cb8a4e-b2ab-4dec-b911-f8658e2f1c6f

staskus commented:
@southp, could you take a look at this draft PR when you have time? Most importantly if the whole structure for creating a new page makes sense.

We decided to create a separate page for Jetpack App Plan Selection during Site Creation so we wouldn't need to make too many alternations to existing flows and components. Outside a header and loading, most of the functionality is contained within `PlanFeaturesMain` component.

I'm not sure what are the rules behind the placement and naming of new pages. I put it into `.com/jetpack-app-plans` but I could restructure or rename it in a preferred way.  

Thanks!

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation

staskus committed:
Renaming plans-grid path

staskus commented:
>  I don't have time to look deeper into the implementation details, but structural-wise you've done it correctly, and you've also made use of the intent system the right way to present the curated set of plans :)

Thanks for taking a look at it, @southp  🙇 . We're not developing much (or at all) in Calypso repository, so it's good to know that it's done correctly structurally. Since this page will be only shown through the app, we can own any styling or navigation related issues.

staskus committed:
Simplify redirect by passing only plan product id

staskus committed:
Remove redundant flow name

staskus committed:
Set a lighter web view color for Jetpack App Plans to match other plans page shown on the app

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation

staskus commented:
@southp do you think you would have the capacity to take a look at it one more time for approval?

@enejb maybe you would be available to review these changes? This is a separate page to display plans on the jetpack app, it shouldn't affect Calypso directly. 🙇

jdc91 commented:
Hi @staskus 👋 !
Nice work setting this up on Calypso on your first attempt. While everything looks great I think we can explore a better approach when considering this change.

https://github.com/Automattic/wp-calypso/pull/82135/commits/20b2a8ef27647a30c655bc474978635f779a8ef8

Looks like you need to bypass the default layout of calypso and simply load the plans grid without all the surrounding layout decoration available by default on all calypso related pages.

This layout decoration happens here

https://github.com/Automattic/wp-calypso/blob/e92268a32a6ec183195c39c8046f37e038d88386/client/controller/shared.js#L7-L47

This is via the `makeLayout` middleware call here
https://github.com/Automattic/wp-calypso/blob/6f08ea2be46cadff7b7a3b8664a26df4eb54dda8/client/jetpack-app-plans/index.js#L6-L15

What you can do is write your own `makeLayoutMiddleware` something like `makeJetpackLayoutMiddleware` so that it can be used in future jetpack work related to this type of tasks.

Tagging @tyxla for possible feedback about this approach.

staskus committed:
Create a middleware for Jetpack App pages layout

Jetpack App pages layout do not require all the outer layouts of Calypso. We can omit them by creating a separate middleware that only displays the required component.

staskus committed:
Change jetpack-app-plans to jetpack-app/plans route

staskus committed:
Remove redundant style

staskus commented:
> Hi @staskus 👋 !
Nice work setting this up on Calypso on your first attempt. While everything looks great I think we can explore a better approach when considering this change.

Thanks @jdc91 for the review! Yes, this is the first attempt but it's likely we'll have more and more integrations going forward so it's good to raise the competence and understanding.

> What you can do is write your own makeLayoutMiddleware something like makeJetpackLayoutMiddleware so that it can be used in future jetpack work related to these type of tasks.

Okay, got it. Yes, I admit I haven't located the exact place where the decorations are coming, thanks for pointing it out!

- I created a middleware https://github.com/Automattic/wp-calypso/pull/82135/commits/a06919a29bb2bdf90f18ce001fc09ccc93d30dee which doesn't wrap our component in the `Calypso` layouts but keeps the other parameters I found on similar middleware components
- Given one of the intentions of this middleware is reusability for other Jetpack App cases, I adjusted the routing and folder structure to be `jetpack-app/plans` rather than `jetpack-app-plans`, so we would have a consistent structure if we add more standalone features. https://github.com/Automattic/wp-calypso/pull/82135/commits/b6eac81dc31a9e0cf71f42a8e6d68eba8f40f851

jdc91 commented on client/jetpack-app/plans/main.jsx:
```diff
@@ -0,0 +1,109 @@
+	</div>
+);
+
+const Plans = ( { paidDomainName, onUpgradeClick, handleRedirect } ) => (
+	<PlansFeaturesMain
+		paidDomainName={ paidDomainName }
+		intent="plans-jetpack-app-site-creation"
+		isInSignup={ true }
+		intervalType="yearly"
+		onUpgradeClick={ onUpgradeClick }
+		plansWithScroll={ false }
+		removePaidDomain={ handleRedirect }
+		hidePlanTypeSelector={ true }
+		hidePlansFeatureComparison={ true }
+	/>
+);
+
+const Loading = () => (
+	<div className="plans__loading">
+		<LoadingEllipsis active />
+	</div>
+);
```
Unnecessary sub component. Don't be afraid of a large JSX tree. There is no reason to move parts of the tree into smaller components unless there is some data related complexity or until you sense the need to reuse a component elsewhere. Just keep the tree inline as much as possible. This way it's much easier to reason about the code without haven't to click through multiple times to reach the changed piece of code.

jdc91 commented on client/jetpack-app/plans/main.jsx:
```diff
@@ -0,0 +1,109 @@
+	<div className="plans__header">
+		<FormattedHeader
+			brandFont
+			headerText={ translate( 'Choose the perfect plan' ) }
```
You need to use the `useTranslate` hook so that the layout is sensitive to local changes once the app has loaded.

ddc22 replied:
You need to use the `useTranslate` hook so that the layout is sensitive to local changes once the app has loaded.

ddc22 commented on client/jetpack-app/plans/main.jsx:
```diff
@@ -0,0 +1,109 @@
+	<div className="plans__loading">
+		<LoadingEllipsis active />
+	</div>
+);
```
Unnecessary sub component!
 Don't be afraid of a large JSX tree. There is no reason to move parts of the tree into smaller components unless there is some data related complexity or until you sense the need to reuse a component elsewhere. This is not a hard and fast rule but, do try to keep the tree inline as much as possible. 
 
 This way it's much easier to reason about the code without haven't to click through multiple components to reach the changed piece of code.

staskus replied:
Thanks, good comment. We don't want to be extracting extra components for the sake of doing that. Especially if it decreases the readability of the code 👍 .

ddc22 commented on client/jetpack-app/plans/main.jsx:
```diff
@@ -0,0 +1,109 @@
+	const planSlug = useSelector( ( state ) =>
+		getPlanSlug( state, getPlan( PLAN_FREE )?.getProductId() || 0 )
+	);
+	const plansLoaded = Boolean( planSlug );
```
There is a separate selector for this.

https://github.com/Automattic/wp-calypso/blob/29282c17f0b180d148134a8fdaaa74fd5e3b0956/client/state/plans/selectors/plan.ts#L18

staskus replied:
Thanks for letting me know about this!

There's a slight issue. I have a `<QueryPlans />` component that starts loading the Plans. If I use

```
const isPlansListLoading = useSelector( ( state: AppState ) => isRequestingPlans( state ) );
```

within `JetpackAppPlans` component, then there's a constant switching between `<QueryPlans />` triggering loading state, and then `isPlansListLoading` turning to false making the component to re-render triggering `<QueryPlans />` again.

I'm not sure if it's worth extracting `<QueryPlans />` to a parent component to avoid this? 🤔  

For now, I left the solution as it is, especially given `plansLoaded` is used in other Plans-related components.

staskus committed:
Remove Loading and Plans subcomponents to improve readability

staskus committed:
Use useTranslate hook

ddc22 commented on client/jetpack-app/plans/style.scss:
```diff
@@ -0,0 +1,33 @@
+@import "calypso/my-sites/plans-grid/media-queries";
+
+body {
+	background-color: #fdfdfd;
```
I there a CSS variable we can use for this?
https://wpcalypso.wordpress.com/devdocs/docs/color.md

Maybe  --studio-gray-0? (#f6f7f7)

staskus replied:
Good question. This one is a bit too dark, unfortunately. I'm using the color to fit the Jetpack app colors. However, if this would be a blocker we could live with that for now 👍

tyxla replied:
In any case, we shouldn't be declaring colors like this and we should use variables and reuse them throughout the codebase. That being said, is there an existing variable you could reuse here instead of declaring a new color inline?

staskus replied:
Yes, I ended up setting the color as `--studio-white`. This is what works for us.

ddc22 commented on client/jetpack-app/page-middleware/layout.jsx:
```diff
@@ -0,0 +1,53 @@
+import { QueryClientProvider } from '@tanstack/react-query';
+import { Provider as ReduxProvider } from 'react-redux';
```
Can you try reverting all **new files** to tsx? Did you try this before? Were there any challenges that you encountered?

staskus replied:
Okay, I can try to convert them all to `tsx`. I haven't tried it before 👍

ddc22 reviewed (changes requested):
Nice work!
I have some changes that you can work on, but I think we are very close :)

staskus committed:
Convert JS to TS

staskus commented:
> Nice work!
I have some changes that you can work on, but I think we are very close :)

@jdc91, thank you for the review and valuable comments! 🙇 

I addressed most of the comments. I explained what I couldn't make work next to the comments.

tyxla commented on client/jetpack-app/page-middleware/layout.tsx:
```diff
@@ -0,0 +1,61 @@
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
+import { Context } from 'page';
```
If we're using an import only for typing purposes, I'd recommend using `import type` instead. This applies to multiple other type imports throughout the PR.

tyxla commented on client/jetpack-app/page-middleware/layout.tsx:
```diff
@@ -0,0 +1,61 @@
+	);
+};
+
+export function makeJetpackAppLayoutMiddleware( LayoutComponent: typeof ProviderWrappedLayout ) {
```
Do we need the `makeJetpackAppLayoutMiddleware()` abstraction? Feels like we could just declare the returned middleware function itself and use that directly.

I know @jdc91 suggested that earlier, but it might be an unnecessary abstraction, and we can always add an abstraction when we really need it (see the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle).

staskus replied:
I don't have a strong opinion on this. I don't want Jetpack-app-specific pages to break established conventions and affect unrelated parts of Calypso. 

I started by reusing existing middleware but it required making additional changes around the edges to disable unwanted Calypso layout features. 

Not sure which one is preferable.

tyxla replied:
I'd say this is a minor implementation detail, while I prefer the simpler version, feel free to go with whatever works best for your case.

tyxla commented on client/jetpack-app/plans/index.ts:
```diff
@@ -0,0 +1,16 @@
+
+export default function () {
+	page(
+		'/jetpack-app/plans',
```
Let's make sure we're handling the root route properly, since on production the `/jetpack-app` route will be mapped on nginx to run on Calypso, and we should handle all cases under it properly.

staskus replied:
Got it, thanks!

I made changes in this commit https://github.com/Automattic/wp-calypso/pull/82135/commits/0889d5ca33cfb9fcd8af5f692aea9aa228449fb4. I'm handling the whole `jetpack-app` in `sections.js` and defaulting to `jetpack-app/plans` given we don't have any other pages for now.

tyxla commented on client/jetpack-app/plans/controller.tsx:
```diff
@@ -0,0 +1,12 @@
+import JetpackAppPlans from './main';
+
+export function jetpackAppPlans( context: PageJS.Context, next: () => void ) {
```
I see here we're using `PageJS.Context` without importing the type, while in other places we import the `Context` type. It could use some consistency.

tyxla commented on client/jetpack-app/plans/controller.tsx:
```diff
@@ -0,0 +1,12 @@
+	context.primary = (
+		<JetpackAppPlans
+			paidDomainName={ context.query.paid_domain_name }
+			redirectTo={ context.query.redirect_to }
```
Noting that `redirect_to` query arguments are often subject to open redirect vulnerabilities, so it's something to keep in mind.

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+					</p>
+					<p>
+						{ translate(
+							'You’ll also unlock advanced features that make it easy to build and grow your site.'
```
Are the strings final? I'm asking because when committed, they'll be sent for paid translation to all mag-16 languages (see PCYsg-9nE-p2) .

staskus replied:
I reuse these strings [from another component](https://github.com/Automattic/wp-calypso/blob/1cf56b0bc88ad981494c9e3fa30d26d999ed2320/client/my-sites/plans/main.jsx#L89).

Are they reused or am I required to reuse the whole component in such a scenario?

tyxla replied:
As long as we use the same strings with the same `translate()` arguments, the strings are going to be reused.

tyxla replied:
Of course, if you can reuse an entire component, that might be a good idea if it's cheap.

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+	const handleRedirect = ( planId?: number ) => {
+		if ( redirectTo ) {
+			const cartItemParam = planId ? `?plan_id=${ planId }` : '';
+			window.location.href = `${ redirectTo }${ cartItemParam }`;
```
This looks potentially vulnerable to open redirects as mentioned above.

staskus replied:
Thanks for notifying us about this. Could you share any alternative approaches we could use?

**Update**

I removed `redirect_to` since it's not required for us. What we need to have is a mechanism to notify the app's web view when the plan is selected. For now, I haven't changed approach too much, just used `originalUrl` instead to construct the path https://github.com/Automattic/wp-calypso/pull/82135/commits/ada828f4af6555a012125078644a3fababc70ee9

tyxla replied:
Sounds good 👍

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+
+	const handleRedirect = ( planId?: number ) => {
+		if ( redirectTo ) {
+			const cartItemParam = planId ? `?plan_id=${ planId }` : '';
```
Could be useful to use `addQueryArgs()` here.

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+					<PlansFeaturesMain
+						paidDomainName={ paidDomainName }
+						intent="plans-jetpack-app-site-creation"
+						isInSignup={ true }
```
```suggestion
						isInSignup
```

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+						onUpgradeClick={ onUpgradeClick }
+						plansWithScroll={ false }
+						removePaidDomain={ handleRedirect }
+						hidePlanTypeSelector={ true }
```
```suggestion
						hidePlanTypeSelector
```

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+						plansWithScroll={ false }
+						removePaidDomain={ handleRedirect }
+						hidePlanTypeSelector={ true }
+						hidePlansFeatureComparison={ true }
```
```suggestion
						hidePlansFeatureComparison
```

tyxla commented on client/sections.js:
```diff
@@ -215,6 +215,11 @@ const sections = [
 	},
+	{
+		name: 'jetpack-app-plans',
+		paths: [ '/jetpack-app/plans' ],
```
As mentioned above, we'll need to handle the entire `/jetpack/app` route somewhere. Could be in another section, but it needs to be handled.

tyxla commented on client/sections.js:
```diff
@@ -215,6 +215,11 @@ const sections = [
 	},
+	{
+		name: 'jetpack-app-plans',
+		paths: [ '/jetpack-app/plans' ],
```
This also made me question the route: `/jetpack-app` is a bit of an unusual user-facing route. `app` is not a term we often see in user-facing routes. At the same time, once a route enters production it has to be maintained nearly forever. Has that been considered?

staskus replied:
Our use case is to load a plan selection page from within the app. 

While we were able to reuse `plans/yearly` route for the post-site creation flow on the app, the `start/` flow required too many granular changes in various places to be adapted for the app. That's where we decided to create a separate page reusing `PlanFeaturesMain` component.

> This also made me question the route: /jetpack-app is a bit of an unusual user-facing route. app is not a term we often see in user-facing routes.

It's a very good question. The route is meant to be opened from the app but never directly typed by the user. If this page could be accessed by reusing other routes with certain parameters that would work equally well for us. 

> At the same time, once a route enters production it has to be maintained nearly forever. Has that been considered?

Our expectation is for this route to be maintained for as long as this route is used by the app. If we decide to disable the feature or use another method of Plan purchasing, we could remove this route.

enejb replied:
Maybe we can redirect any user that is coming to this page not from the app?

staskus replied:
I tried this approach and it looks to be working well. I used the approach of similar conditional redirects that I found in the codebase.

If `jetpack-app/*` route is not accessed through the app I'm redirecting home https://github.com/Automattic/wp-calypso/pull/82135/commits/5a7dbaa1eddf9765d643683d9433a3c13e8534c9:

https://github.com/Automattic/wp-calypso/assets/4062343/7141e6b5-1dcc-4ce6-b6a2-85da5a33e0b4

@tyxla what do you think about such an approach? Does it alleviate some of the concerns?

tyxla replied:
Sounds good, thanks for handling it 👍

tyxla commented on client/jetpack-app/plans/style.scss:
```diff
@@ -0,0 +1,33 @@
+			font-size: $font-body;
+			color: var(--color-text-subtle);
+			text-align: center;
+			padding: 0 16px;
```
Ideally, we should use the same spacing as the rest of the app. I realize some parts aren't very consistent, but for newly introduced pieces, we could probably try to reuse an existing variable.

staskus replied:
That's a good point. The only comprehensive global spacing declarations I found were `grid-unit` https://github.com/Automattic/wp-calypso/pull/82135/commits/e1b4a6b2466c343164ccb03e3dacffb0325627a8, I'm using them instead of hardcoded values.

staskus committed:
Use --studio-white for Jetpack App Plans background color

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,113 @@
+					/>
+				</>
+			) : (
+				<div className="plans__loading">
```
Might want to change some classnames to fit our [coding guidelines](https://github.com/Automattic/wp-calypso/blob/trunk/docs/coding-guidelines/css.md#classes). The biggest deal here is that this classname might be in use already and we might need a proper prefix that fits the context, like `jetpack-app__plans-` or `jetpack-app-plans__`.

staskus committed:
Remove explicit { yes } from parameters

staskus committed:
Use import type for imports only for typing purposes

staskus committed:
Use a consistent Context definition

staskus committed:
Use addQueryArgs for building redirection URL

staskus committed:
Use a jetpack-app__ prefix for class names

staskus committed:
Use grid-unit variables for spacing

staskus committed:
Redirect to originalUrl with select plan_id and plan_slug

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation

staskus committed:
Handle an entire jetpack-app/ route

Redirect to jetpack-app/plans when accessing jetpack-app route, since it's the only page of the route for now

staskus committed:
Redirect to home if jetpack-app/ route is not accessed via the app

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,111 @@
+				headerText={ translate( 'Choose the perfect plan' ) }
+				align="center"
+			/>
+			{ paidDomainName ? (
```
One concern here (not necessarily a blocker for this PR): we don't do any domain checks as early as here. 

So if someone inputs a domain that is not a valid domain (like `<script>alert(1)</script>`) or a domain that is not available (if it's registered already for example), we're still going to suggest that the user can get that domain here. 

For example, try with `/jetpack-app/plans?paid_domain_name=wordpress.com`. The user won't be able to purchase the domain because it's not available, but we're still suggesting that they could buy it with our plan.

I think this could be confusing and we should display this message only if the domain is actually available for purchase.

staskus replied:
Yes, that's a fair point. We do the domain selection on the app, and this `paid_domain_name` is passed only for UI configuration. However, it does allow for misleading or invalid configuration of this page.

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,111 @@
+		if ( ! productSlug ) {
+			args = { plan_slug: PLAN_FREE };
+		} else {
+			const plan = getPlan( productSlug ) as Plan;
```
Ideally, there wouldn't be so much need for casting and types would be either inferred or specifically declared. Not a blocker for this PR, just something to keep in mind.

tyxla commented on client/jetpack-app/plans/style.scss:
```diff
@@ -0,0 +1,43 @@
+
+		.formatted-header__title,
+		.formatted-header__subtitle {
+			max-width: unset;
```
Not a blocker, but a good idea for a follow-up: would be great to explore which styles we can more generically adopt for that component so they apply to all mobile devices. That could be a big win: in every other place where the `FormattedHeader` component gets utilized, it would already look well on mobile.

tyxla commented on client/sections.js:
```diff
@@ -215,6 +215,11 @@ const sections = [
+	{
+		name: 'jetpack-app',
+		paths: [ '/jetpack-app' ],
+		module: 'calypso/jetpack-app/',
```
For consistency we could remove the forward slash.

```suggestion
		module: 'calypso/jetpack-app',
```

tyxla commented on client/jetpack-app/index.ts:
```diff
@@ -0,0 +1,17 @@
+
+export default function () {
+	page(
+		'/jetpack-app/plans',
```
If this is not production-ready, we might have to consider wrapping the feature in a feature flag. 

Here's an example how to do that at the `controller` level: 

https://github.com/Automattic/wp-calypso/blob/e1f240e1fb5a4083592a787b3e6a4f4c9f553692/client/jetpack-connect/controller.js#L269

where we declare that feature as enabled in the various environments (each represented by a JSON file under `/config`.

tyxla commented on client/jetpack-app/index.ts:
```diff
@@ -0,0 +1,17 @@
+import page from 'page';
```
Not a blocker for this PR: Would be great to add a README for the `client/jetpack-app` directory that explains its purpose. As a user and as a Calypso developer who sees this for the first time, I can get confused as to what its purpose is, where it's supposed to be accessed, and what it aims to achieve.

tyxla commented on client/jetpack-app/index.ts:
```diff
@@ -0,0 +1,17 @@
+	);
+
+	// Default to /plans page until we have more pages
+	page( '/jetpack-app', '/jetpack-app/plans' );
```
Note that when this lands in production, to access the new `/jetpack-app` route it in staging/production you'll need to request the route to be pointed to Calypso.

Here's an example recent request for another route: pMz3w-ilk-p2

tyxla commented on client/jetpack-app/plans/main.tsx:
```diff
@@ -0,0 +1,111 @@
+			{ plansLoaded ? (
+				<>
+					<Header paidDomainName={ paidDomainName } />
+					<PlansFeaturesMain
```
Minor and not a blocker for this PR: when clicking the "Show all features" link of a plan, that link moves slightly up in the Jetpack App version, while in the .com plans one it stays on the same place: https://wordpress.com/pricing/

tyxla reviewed (approved):
This looks good and works well in my testing 👍 Nice work 🙌 

I've left a few suggestions for follow-ups and a few notes to keep in mind.

Feel free to ship as-is and address the feedback in future PRs based on your project's prioritization.

🚀

staskus committed:
Added README.md explaining jetpack-app/ route

staskus committed:
Remove forward slash from the section path

staskus commented:
I added README.md and noted other improvements in the following task https://github.com/Automattic/wp-calypso/issues/82934. We want to make more UI adjustments so other improvements will fit well as well.

staskus committed:
Merge branch 'trunk' into task/82131-create-plans-page-for-jetpack-app-site-creation
